### PR TITLE
Range-ify std.file.write(), readText() and append()

### PR DESCRIPTION
**Pull request by WalterBright**
Also:
1. combined common code in read() implementation
2. updated Ddoc comments

http://dlang.org/phobos/std_file.html#.write

### CODE REVIEW

File: std/file.d [hunk 1]
```diff
@@ -42,6 +42,18 @@ else version (Posix)
 else
     static assert(false, "Module " ~ .stringof ~ " not implemented for this OS.");
 
+// Character type used for operating system filesystem APIs
+version (Windows)
+{
+    private alias FSChar = wchar;
```
Comment (andralex) [inline std/file.d]: noice, amazing this didn't come up yet

File: std/file.d [hunk 2]
```diff
@@ -215,19 +227,25 @@ void main()
...
 ----
 
+Params:
+    name = string or range of characters representing the file _name
+    upTo = if present, the maximum number of bytes to read
+
 Returns: Untyped array of bytes _read.
 
-Throws: $(D FileException) on error.
+Throws: $(LREF FileException) on error.
```
Comment (andralex) [inline std/file.d]: noice

**Comment by andralex** [comment]
alrighty then

**Comment by CyberShadow** [comment]
OK, so this actually broke code for me.

I have a wrapper that wraps arbitrary file operation functions and makes them atomic (by writing to a temporary file and renaming it onto the target). I use `ParameterTypeTuple` to find which parameter contains the "target path" string, which doesn't work now that the function is a template.

I also had a similar breakage when `std.file.copy` gained a third optional parameter. Since the usual way to create a function wrapper (using `ParameterTypeTuple`) does not account for optional parameters, the parameter became non-optional in the wrapped function, which broke code using it.

It seems to me that to truly avoid all breaking changes, we can't touch any Phobos function signatures at all. This is probably too drastic, so I think we need to establish some conventions of what breaking changes are acceptable.

**Comment by CyberShadow** [comment]
@WalterBright @andralex ping

I filed an issue so this doesn't get lost:
https://issues.dlang.org/show_bug.cgi?id=14742

**Pull request merged**